_posts/2019-09-19-sorting-a-table-with-live-view-live-links.md

```diff
@@ -0,0 +1,197 @@
+  case sort_by do
+    sort_by
+    when sort_by in ~w(name) ->
+      {:noreply, assign(socket, cohorts: sort_cohorts(socket.assigns.cohorts, sort_by))}
+
+    _ ->
+      {:noreply, socket}
+  end
+end
+
+
+def handle_params(_params, _uri, socket) do
+  {:noreply, socket}
+end
+
+def sort_cohorts(cohort, "name") do
+  Enum.sort_by(cohorts, fn cohort -> cohort.name end)
+end
+```
+
+Note that we've included a "catch-all" version of the `handle_params/3` function that will be invoked if someone navigates to `/cohorts` and includes query params that do not match the `"sort_by"` param that we care about. If our live view recieves such a request, it will not update state.
+
+Now, when a user clicks the `"Name"` live link, two things will happen:
+
+* The browser's pushState API will be leveraged to change the URL to `/cohorts?sort_by=name`
+* Our already-mounted live view's `handle_params/3` function will be invoked with the params `%{"sort_by" => "name"}`
+
+Our `handle_params/3` function will then sort the cohorts stored `socket.assigns` by cohort name and update the socket state with the sorted list. The template will therefore re-render with the sorted list.
+
+Since `handle_params/3` is _also_ called after `mount/2`, we have therefore allowed a user to navigate directly to `/cohorts?sort_by=name` via their browser and see the live view render with a table of cohorts already sorted by name. And just like that we've enabled users to share links to sorted table views with zero additional lines of code!
+
+## More Sorting!
+
+Now that our "sort by name" feature is up and running, let's add the remaining live links to allow users to sort by the other attributes we listed earlier: campus, start date and status.
+
+First, we'll make each of these table headers into a live link:
+
+```html
+<table>
+  <th><%= live_link "Name", to: Routes.live_path(@socket, CourseConductorWeb.CohortsLive, %{sort_by: "name"}) %></th>
+  <th><%= live_link "Campus", to: Routes.live_path(@socket, CourseConductorWeb.CohortsLive, %{sort_by: "campus"}) %></th>
+  <th><%= live_link "Start Date", to: Routes.live_path(@socket, CourseConductorWeb.CohortsLive, %{sort_by: "start_date"}) %></th>
+  <th><%= live_link "Status", to: Routes.live_path(@socket, CourseConductorWeb.CohortsLive, %{sort_by: "status"}) %></th>
+</table>
+```
+
+And we'll build out our `handle_params/3` function to operate on params describing a sort by any of these attributes:
+
+```elixir
+def handle_params(%{"sort_by" => sort_by}, _uri, socket) do
+  case sort_by do
+    sort_by
+    when sort_by in ~w(name course_offering campus start_date end_date lms_cohort_status) ->
+      {:noreply, assign(socket, cohorts: sort_cohorts(socket.assigns.cohorts, sort_by))}
+
+    _ ->
+      {:noreply, socket}
+  end
+end
+```
+
+Here, we've added a check to see if the `sort_by` attribute is included in our list of sortable attributes.
+
+```elixir
+when sort_by in ~w(name campus start_date status)
```

Comment:
i would add all the `sort_by` attributes for clarity:
`when sort_by in ~w(name course_offering campus start_date end_date lms_cohort_status)`